ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Reader.hs

```diff
@@ -0,0 +1,500 @@
+     -- ^ The current chain fragment
+  -> (ReaderRollState blk -> STM m ())
+     -- ^ How to save the updated 'ReaderRollState'
+  -> STM m (Maybe (ChainUpdate blk (Header blk)))
+instructionSTM rollState curChain saveRollState =
+    assert (invariant curChain) $ case rollState of
+      RollForwardFrom pt ->
+        case AF.successorBlock (castPoint pt) curChain of
+          -- There is no successor block because the reader is at the head
+          Nothing  -> return Nothing
+          Just hdr -> do
+            saveRollState $ RollForwardFrom $ headerPoint hdr
+            return $ Just $ AddBlock hdr
+      RollBackTo      pt -> do
+        saveRollState $ RollForwardFrom pt
+        return $ Just $ RollBack pt
+  where
+    invariant =
+      AF.withinFragmentBounds (castPoint (readerRollStatePoint rollState))
+
+forward
+  :: forall m blk.
+     ( MonadSTM   m
+     , MonadCatch m
+     , HasHeader blk
+     , HasHeader (Header blk)
+     , HasCallStack
+     )
+  => ChainDbEnv m blk
+  -> TVar m (ReaderState m blk)
+  -> [Point blk]
+  -> m (Maybe (Point blk))
+forward CDB{..} varReader = \pts -> do
+    -- The current state of the reader doesn't matter, the given @pts@ could
+    -- be in the current chain fragment or in the ImmutableDB.
+
+    -- NOTE: we don't use 'Query.getCurrentChain', which only returns the last
+    -- @k@ headers, because we want to see the headers that have not yet been
+    -- written to the ImmutableDB too.
+    curChain         <- atomically $ readTVar cdbChain
+    slotNoAtImmDBTip <- ImmDB.getSlotNoAtTip cdbImmDB
+    findFirstPointOnChain curChain slotNoAtImmDBTip pts
+  where
+    findFirstPointOnChain
+      :: HasCallStack
+      => AnchoredFragment (Header blk)
+      -> SlotNo
+      -> [Point blk]
+      -> m (Maybe (Point blk))
+    findFirstPointOnChain curChain slotNoAtImmDBTip = \case
+      []     -> return Nothing
+      pt:pts
+        | AF.withinFragmentBounds (castPoint pt) curChain
+        -> do
+          -- It's in the in-memory chain fragment
+          atomically $ writeTVar varReader $ ReaderInMem $ RollBackTo pt
+          return $ Just pt
+
+        | pointSlot pt > slotNoAtImmDBTip
+          -- The point is after the tip of the ImmutableDB, so don't look in
+          -- the ImmutableDB and skip the point.
+        -> findFirstPointOnChain curChain slotNoAtImmDBTip pts
+
+        | pt == genesisPoint
+          -- We know the point is not in the in-memory fragment (not even as
+          -- the anchor point). This means that the ImmutableDB is not empty
+          -- (@AF.anchorPoint curChain /= genesisPoint@).
+        -> do
+          immIt <- ImmDB.streamBlocksAfter cdbImmDB genesisPoint
+          let readerState = ReaderInImmDB (RollBackTo pt) immIt
+          atomically $ writeTVar varReader readerState
+          return $ Just pt
+
+        | otherwise
+        -> do
+          inImmDB <- isJust <$> ImmDB.getBlockWithPoint cdbImmDB pt
```

Comment:
Add something like `ImmDB.pointMember :: Point blk -> m Bool`

Reply:
Combine this case with the `genesisPoint` case: combine the membership test with (`== genesisPoint`)

Reply:
Unfortunately we cannot implement `ImmDB.pointMember :: Point blk -> m Bool` more efficiently than the current approach, because we need to read the block in order to figure out its hash (Point = hash + slot), as the ImmutableDB is indexed by slots.